♻️ Plugin harmonisation

imqdee opened the pull request:
The first steps of issue #5 (excluding the last two)

chabroA commented on src/plugins/telegram/index.ts:
```diff
@@ -1,45 +1,70 @@
-        chat_id: chatId[i],
+  async notify(message: string, targets: string[]): Promise<void> {
+    const endpoint = `${this.#BASE_URL}/sendMessage`;
+    const chatId = await this.#getUserId(targets);
```
I would rename `chatId` to `chatIds` to avoid confusion since this variable represents a collection (list) of chat ids

```suggestion
    const chatIds = await this.#getUserId(targets);
```

chabroA commented on src/plugins/telegram/index.ts:
```diff
@@ -1,45 +1,70 @@
+    const chatId = await this.#getUserId(targets);
+
+    // create an iterable of promises
+    const requests = chatId.map((id) =>
```
```suggestion
    const requests = chatIds.map((id) =>
```

chabroA commented on src/plugins/telegram/index.ts:
```diff
@@ -1,45 +1,70 @@
+    // throw an error if users cannot be found (meaning targetted users didn't start a conversation with the bot)
+    // *OPINIONATED*: we only throw an error if all of the targetted users has not started a conversation with the bot
+    if (UserIds.length === 0)
+      this.throwError('➡️ User has not started conversation with user');
```
This message is a bit ambiguous, it sounds like the first and last "user" are not the same (which is probably the case 🤔)
According to the comment, this error is thrown when the user has not started a conversation with the bot.

```suggestion
      this.throwError('➡️ User has not started conversation with bot');
```

imqdee committed:
♻️ massive plugin refactoring

- Create a plugin interface to unify the plugins
- Create the default plugin class that contains utils functions
- Rewrite all the plugins to match the plugin interface
- Drop the email plugin for the moment (would be rewritten later)
- Improve all the types present in the application
- Extend eslint configuration allowing namespace declaration

imqdee committed:
🚚 rename plugins to connectors

imqdee committed:
➕ add missing ts-node dependency

Co-authored-by: Alexandre Chabrolin <9203826+chabroA@users.noreply.github.com>

imqdee committed:
🐛 Only up the server once database is connected

- Await until the connection with the database is etablished before up the server
- Move the server logic into a dedicated server/ folder
- Create an index file at the root of src/ that loads the server configuration and up it

imqdee committed:
📝 add important comments and improve .env.example

- Improve .env.example by giving more guidance for the developer
- Add a step in the installation guide to create the required .env file
- Add comments to current connectors that explain current weaknesses of
the implementations and explain how to fix them. This is a serious todo list.
